### ISSUE TITLE
Recently open files palette

### ISSUE THREAD

**Issue by rtorr** [zed-industries/zed#4663]
### Check for existing issues

- [X] Completed

### Describe the feature

In other editors (vscode/intellij), they provide a palette with recently open files. You can either fuzzy search or go through the list manually with arrow keys (I usually set a different keybinding for selecting which file). When you arrow down, as the selection changes, the editor will update with that selected file. Hitting enter will dismiss the palette.

I do this because I like to hide editor tabs (another feature request for another time?) and only select recently opened files through this pane.

To demo this flow in vscode, these are the keybindings I use and the commands.

```json
  {
    "key": "cmd+e",
    "command": "workbench.action.openPreviousEditorFromHistory"
  },
  {
    "key": "cmd+e",
    "command": "workbench.action.quickOpenNavigateNext",
    "when": "inQuickOpen"
  },
  ```

**Comment by mikayla-maki** [comment]
Is there any difference between this and the existing tab switcher implementation?

**Comment by gatspy** [comment]
I hope that folders and files can be dragged into panel like the sublime, and the next startup  still keep them intact.

**Comment by nlopin** [comment]
One thing that keeps triggering me is when I open recent files `cmd + p`, the current file is always first and selected. I'd expect the previous file to be first and the current is not shown at all. This way you can easily jump between the last two files by tapping `cmd + p` and then `enter`

**Comment by handlebauer** [comment]
> This way you can easily jump between the last two files by tapping `cmd + p` and then `enter`

 after searching far and wide the consensus seems to be that `cmd-p p` (that's a second `p` while still holding `cmd`)  is what's intended as zed's best alternative. noticeably, this approach requires the same number of keystrokes as `cmd-p` and then `enter`, but after trying to adapt for the last couple months., it still feels nowhere near as natural 😔.

**Comment by tychedelia** [comment]
Yeah, I stopped using Zed because of this.

**Comment by injust** [comment]
> One thing that keeps triggering me is when I open recent files `cmd + p`, the current file is always first and selected. I'd expect the previous file to be first and the current is not shown at all. This way you can easily jump between the last two files by tapping `cmd + p` and then `enter`

FWIW, this is this same behaviour as VS Code.

**Comment by nlopin** [comment]
> FWIW, this is this same behaviour as VS Code

Not sure it proves anything. I use Webstorm. There is a special hotkey for recent files and it always select the previous one, not the current by default

![CleanShot 2024-11-05 at 13 02 22@2x](https://github.com/user-attachments/assets/c36eff6a-8235-4843-b8fa-81854b9d5ba1)

**Comment by asilvadesigns** [comment]
any traction on this?

**Comment by amberpixels** [comment]
+1 for point not to have default cursor on current file in the list of "Project files" (Cmd+P).
What's the chance of user keeping the same file if user already pressed "Cmd+P". The chance is 0%. Then why keep it as a default option?

**Comment by ns-hferreirogarcia** [comment]
I think there might have been quite a lot of progress since this issue was opened. Regarding https://github.com/zed-industries/zed/issues/4663#issuecomment-1908629469 and https://github.com/zed-industries/zed/issues/4663#issuecomment-2390661375, you can just press `cmd+p+p` (that is, press `p` twice and then release both keys).

**Comment by hferreiro** [comment]
I think there might have been quite a lot of progress since this issue was opened. Regarding https://github.com/zed-industries/zed/issues/4663#issuecomment-1908629469 and https://github.com/zed-industries/zed/issues/4663#issuecomment-2390661375, you can just press `cmd+p+p` (that is, press `p` twice and then release both keys).

**Comment by adrian-goe** [comment]
so `cmd+p` is a light version of that, but can anyone tell me what the actual action is, that is being triggered so I can rebind it to `cmd+e`? Can't find it

**Comment by lucamtudor** [comment]
> so `cmd+p` is a light version of that, but can anyone tell me what the actual action is, that is being triggered so I can rebind it to `cmd+e`? Can't find it

```
    "bindings": {
      // open recent file popup
      "cmd-e": "file_finder::Toggle"
    }

```

**Comment by lucamtudor** [comment]
For what is worth, this is my config for VS Code to really emulate JetBrains IDEs behaviour:

```jsonc
  {
    // Makes this identical to Intellij IDEA!
    // It changes the keybinding from IntelliJ IDEA Keybindings:
    // Quick open panel, when it appears, already has selected the previous editor,
    // aka the second row, aka you save a ↓ keystroke.
    "key": "cmd+e",
    "command": "workbench.action.quickOpenPreviousEditor",
    "when": "!inQuickOpen",
    "intellij": "Recent files popup"
  },
  {
    "key": "cmd+shift+e",
    "command": "workbench.action.quickOpen",
    "when": "!inQuickOpen",
    "intellij": "Recent files popup"
  },
  {
    "key": "cmd+e",
    "command": "workbench.action.quickOpenNavigateNextInRecentFilesPicker",
    "when": "inQuickOpen"
  },
  {
    "key": "cmd+shift+e",
    "command": "workbench.action.quickOpenNavigatePreviousInRecentFilesPicker",
    "when": "inQuickOpen"
  }
```

Zed is amazing, I reaaally wished it has these small config flavours like VSCode so I can uninstall that slowshow 😄 .

**Comment by MrBns** [comment]
actually, I stopped using zed because of this issue. my most used shortcut.  
its not about shortcut. its about feature. when you hit shortcut for **Recent Files**, in the list previous file should be appear at the top. but instead Zed show current files at the top.

**Comment by KaKi87** [comment]
@maxdeviant @danilo-leal @ConradIrwin Any news on this please ? :pray:

**Comment by j3g** [comment]
besides the complaint that the currently opened file is showing as first result, the cmd+p is the best solution. It shows you the recently opened files,  *AND* if you start typing it will fuzzy search your project files.

**Comment by KaKi87** [comment]
> is the best solution. It shows you the recently opened files

Only within a currently opened project. If there's no project open and/or if the file isn't part of any project, then it will not be found in this list.

**Comment by injust** [comment]
> > cmd+p is the best solution. It shows you the recently opened files
> 
> Only within a currently opened project. If there's no project open and/or if the file isn't part of any project, then it will not be found in this list.

Projects are basically the expected use case in Zed though, and it would require a huge refactor to change this.

**Comment by KaKi87** [comment]
Zed is regularly compared against Sublime Text and VSCode. Well, they both have this feature.

**Comment by grian32** [comment]
^ Wanted to switch to zed for go and quite like it but unfortunately the lack of this is somewhat of a dealbreaker for me compared to goland. File finder kind of works until you have to use a monorepo and you look for main.go and you end up with 5 files.

**Comment by mrbeiertz** [comment]
I've been back and forth between PyCharm Pro and Zed for the last 6-12 months. There's always something missing that makes me go back to PyCharm. The absence of this feature is HUGE. Is there even a way to implement it manually with Key Bindings?

**Comment by tudorluca** [comment]
> I've been back and forth between PyCharm Pro and Zed for the last 6-12 months. There's always something missing that makes me go back to PyCharm. The absence of this feature is HUGE. Is there even a way to implement it manually with Key Bindings?

Same story here. 

I'm in Intellij IDEA/Android Studio most of my time and always have Zed open as my go to editor of thousands of thousands of lines of text (at least until Swift & Kotlin LSP are on par). CMD-E is the only thing that still trips me over every dang time! 

How about a $500 bounty for this? 🛠️

**Comment by somtooo** [comment]
Another crucial detail missing is the search has to be bounded. with cmd p + p all thats now needed is bounded search rather than a project wide search.  so it has to be a new modal with just that bounded search feature.

**Comment by somtooo** [comment]
> > I've been back and forth between PyCharm Pro and Zed for the last 6-12 months. There's always something missing that makes me go back to PyCharm. The absence of this feature is HUGE. Is there even a way to implement it manually with Key Bindings?
> 
> Same story here.
> 
> I'm in Intellij IDEA/Android Studio most of my time and always have Zed open as my go to editor of thousands of thousands of lines of text (at least until Swift & Kotlin LSP are on par). CMD-E is the only thing that still trips me over every dang time!
> 
> How about a $500 bounty for this? 🛠️

i can add $200 so $700 bounty

**Comment by somtooo** [comment]
> besides the complaint that the currently opened file is showing as first result, the cmd+p is the best solution. It shows you the recently opened files, _AND_ if you start typing it will fuzzy search your project files.

no you dont want to fuzzy search all your project files e.g imagine you have thousands of files called index.tsx you want the relevant index.tsx to show

**Comment by KaKi87** [comment]
That's not a problem actually, look at JetBrains IDEs as mentioned for example, these deal with this perfectly.

**Comment by somtooo** [comment]
> That's not a problem actually, look at JetBrains IDEs as mentioned for example, these deal with this perfectly.

Yeah IntelliJ in the recent file modal will search just the 50 items

**Comment by KaKi87** [comment]
But also, it allows searching `<parent directory name>/index.tsx`.

**Comment by somtooo** [comment]
> But also, it allows searching `<parent directory name>/index.tsx`.

What you mean? Are you saying it searches all files in the recent files modal?

**Comment by KaKi87** [comment]
Well, I guess I'm not sure what you mean by "*recent* files modal", personally I was talking about the [Search Everywhere](https://blog.jetbrains.com/idea/2020/05/when-the-shift-hits-the-fan-search-everywhere/) feature, that you can open by pressing the Shift key twice.

And that feature doesn't only allow specifically browsing *recent* files, but *all*.

**Comment by somtooo** [comment]
> Well, I guess I'm not sure what you mean by "_recent_ files modal", personally I was talking about the [Search Everywhere](https://blog.jetbrains.com/idea/2020/05/when-the-shift-hits-the-fan-search-everywhere/) feature, that you can open by pressing the Shift key twice.
> 
> And that feature doesn't only allow specifically browsing _recent_ files, but _all_.

Ohh yeah that’s correct but issue is focused on recent files palette. You don’t want to search all files in there will take you longer to find a file. Nevertheless I think this was solved with :ls

**Comment by handlebauer** [comment]
> > How about a $500 bounty for this? 🛠️
> 
> i can add $200 so $700 bounty

i'll add 300$ - bounty's up to $1,000

**Comment by somtooo** [comment]
> > > How about a $500 bounty for this? 🛠️
> > 
> > 
> > i can add $200 so $700 bounty
> 
> i'll add 300$ - bounty's up to $1,000

Lmao I think with :ls this is somewhat solved

**Comment by tudorluca** [comment]
> > > > How about a $500 bounty for this? 🛠️
> > > 
> > > 
> > > i can add $200 so $700 bounty
> > 
> > 
> > i'll add 300$ - bounty's up to $1,000
> 
> Lmao I think with :ls this is somewhat solved

To clarify, the Zed keymap for Vim's `:ls`, aka Show all buffers,  is `"tab_switcher::ToggleAll"`, so your keymap.json would have smth like this:

```
[
  {
    "bindings": {

      // open recent file popup
      "cmd-e": "tab_switcher::ToggleAll",
      
      ...
    }
  }
]
```

**Comment by KaKi87** [comment]
Hi,

In that case, I would like to suggest shifting the issue to non-project files.

Thanks

**Comment by hyperknot** [comment]
This is one of the features blocking me from coming from JetBrains. Cmd + E is so prominent there that they put it on their home screen.

**Comment by MrBns** [comment]
> 2\. **Previous file pre-selected** — The setting `file_finder.skip_focus_for_active_in_search` (enabled by default) makes the file finder skip the currently open file and pre-select the previous one. This means `cmd+p → Enter` immediately switches to your last file — the core workflow requested here.

I doesn't work. I got same result either its turned on or off.

**Comment by javid-gulamaliyev** [comment]
> This issue appears to be addressed by existing functionality in Zed:
> 
> 1. **Recent files in the file finder** — `cmd+p` (or `ctrl+p`) already shows recently opened files, including files that have been closed, ordered by recency.
> 
> I think this can be closed.

In JetBrains, the power of the feature comes from the fact that you can search within recent files. Using `cmd + p` starts a new file search, which defeats the purpose of having a quick access list. Additionally, JetBrains allows a one click "filter" add-on to only show edited files, which is beyond useful.

> 3. Tab switcher — tab_switcher::Toggle (ctrl+tab by default) provides MRU-ordered switching across open tabs with preview on navigation.

This isn't even relevant to be honest, it assumed that files are still open.

**Comment by yidasanqian** [comment]
I’d like to add a related use case here: it would be very helpful if "recent files" included files opened outside the current workspace/project as well.

My main use cases are:
- files opened directly from Finder
- config files outside a project
- temporary notes
- one-off edits in unrelated directories

Right now, workspace/project-scoped recent files are useful, but they do not help much with standalone files that are not part of the active workspace.

A global recent files list would make it much easier to jump back to these files across unrelated tasks.

**Comment by msmiller** [comment]
Just want to add another +1 for this feature. I'm often working in a wide range of areas in a Rails project (views, models, controllers, components, policy). Having 30 or more files open at once it not uncommon. It's quicker to be able to see the list on the right side and go right to what I want than to have to use gestures to go find what I want. The project navigator doesn't help because things are in multiple directories and looking in there once again means gestures and taking focus onto finding what I want.

Yes, the open-files tab does the same thing, but it requires more work and also shifting concentration. It's the difference between a HUD and looking down at your dashboard to read the speedo.

**Comment by EaglePPP** [comment]
Either it should be like other IDE or text editors let people "open recent files" in "files menu" or put that into the command palatte.

I love this programm and use it every day, its snappy and snappy and snappy, but without this function it will always be a half finished project to me. and thats sad bc i have to use Everything instead to search for already once opened documents which have weird names regularly.

**Comment by anveshasaraf** [comment]
Picking this up — opened a first PR (#62188) with a dedicated, searchable "Recently Opened Files" palette (separate from the file finder's history mode), listing files from navigation history newest-first with substring filtering.

That PR is scoped to in-memory history only (matches how history already works elsewhere in Zed today). A follow-up PR will add persistence so recent files survive a restart, since that seems to be the biggest recurring ask in this thread.

Left two open questions on the PR itself (default keybinding, and whether a separate modal vs. a file-finder mode is preferred) — would appreciate a maintainer's take on either.

**Comment by anveshasaraf** [comment]
Follow-up PR is up: #62247, adding persistence so recent files survive a restart (depends on #62188 above — that one's the palette itself, this one just adds the durable history on top).
